src/libp2p_ipc/libp2p_ipc.capnp

```diff
@@ -0,0 +1,171 @@
+  privateKey @0 :Text;
+  publicKey @1 :Text;
+  peerId @2 :Text;
+}
+
+struct GatingConfig {
+  bannedIPs @0 :List(Text);
+  bannedPeerIDs @1 :List(Text);
+  trustedIPs @2 :List(Text);
+  trustedPeerIDs @3 :List(Text);
+  isolate @4 :Bool;
+}
+
+struct Libp2pConfig {
+  statedir @0 :Text;
+  privateKey @1 :Text;
+  networkId @2 :Text;
+  listenOn @3 :List(Text);
+  metricsPort @4 :Int16;
+  externalMultiaddr @5 :Text;
+  unsafeNoTrustIP @6 :Bool;
+  flood @7 :Bool;
+  peerExchange @8 :Bool;
+  directPeers @9 :List(Text);
+  seedPeers @10 :List(Text);
+  gatingConfig @11 :GatingConfig;
+  maxConnections @12 :UInt32;
+  validationQueueSize @13 :UInt32;
+  minaPeerExchange @14 :Bool;
+}
+
+struct Publish {
+  topic @0 :Text;
+  data @1 :Data;
+}
+
+struct Subscribe {
+  topic @0 :Text;
+  subscriptionId @1 :UInt64;
+}
+
+struct OutgoingStream {
+  peer @0 :Text;
+  protocolId @1 :Text;
+}
+
+struct Stream {
+  id @0 :UInt64;
+  peer @1 :PeerInfo;
+}
+
+struct NewPeer {
+  multiaddr @0 :Multiaddr;
+  isSeed @1 :Bool;
+}
+
+interface Libp2pHelper {
+  configure @0 (config :Msg(Libp2pConfig)) -> (result :VoidResult);
+  setGatingConfig @1 (gatingConfig :Msg(GatingConfig)) -> (result :VoidResult);
+
+  listen @2 (iface :Msg(Text)) -> (result :Result(List(Multiaddr)));
+  getListeningAddrs @3 (void :VoidMsg) -> (result :Result(List(Multiaddr)));
+  beginAdvertising @4 () -> (result :VoidResult);
+
+  addPeer @5 (peer :Msg(NewPeer)) -> (result :Result(NewPeer));
+  listPeers @6 (void :VoidMsg) -> (result :Result(List(PeerInfo)));
+
+  generateKeypair @7 (void :VoidMsg) -> (result :Result(Libp2pKeypair));
+
+  publish @8 (publish :Msg(Publish)) -> (result :VoidResult);
+  subscribe @9 (subscribe :Msg(Subscribe)) -> (result :VoidResult);
+  # do we use this anymore?
+  unsubscribe @10 (subscriptionId :Msg(SubscriptionId)) -> (result :VoidResult);
+  addStreamHandler @11 (protocol :Msg(Text)) -> (result :VoidResult);
+
+  # do we use this anymore?
+  removeStreamHandler @12 (protocol :Msg(Text)) -> (result :VoidResult);
+  openStream @13 (stream :Msg(OutgoingStream)) -> (result :Result(Stream));
+  closeStream @14 (streamId :Msg(StreamId)) -> (result :VoidResult);
+  resetStream @15 (streamId :Msg(StreamId)) -> (result :VoidResult);
+  sendStream @16 (msg :Msg(StreamMessage)) -> (result :VoidResult);
+
+  setNodeStatus @17 (status :Msg(Data)) -> (result :VoidResult);
+  getPeerNodeStatus @18 (peer :Msg(Multiaddr)) -> (result :Result(Data));
+}
+
+
+interface Validation {
+  validate @0 (validation :Msg(Validation)) -> (result :VoidResult);
+}
+
+struct GossipMessage {
+  sender @0 :PeerInfo;
+  seenAt @1 :UInt64;
+  expiration @2 :UInt64;
+  subscriptionId @3 :UInt64;
+  data @4 :Data;
+  validation @5 :Validation;
+}
+
+struct IncomingStream {
+  id @0 :UInt64;
+  peer @1 :PeerInfo;
+  protocol @2 :Text;
+}
+
+struct StreamLost {
+  id @0 :UInt64;
+  reason @1 :Text;
+}
+
+struct StreamMessage {
+  id @0 :UInt64;
+  data @1 :Data;
+}
+
+interface Daemon {
```

Comment:
What's the purpose of Daemon interface?

Reply:
These are the messages the libp2p_helper sends to the daemon. In the current Go description of the IPC protocol, they are referred to as "upcalls".

Reply:
A syntax question: is `gossipReceived @1 (validation :Msg(GossipMessage));` a description of procedure? Is it roughly an equivalent of `gossipReceived @1 (validation :Msg(GossipMessage)) -> void;`?

Reply:
Correct, it is shorthand for a procedure with no return. I believe both definitions your provided are equivalent.